tools/partition_graph_lp.py

```diff
@@ -149,15 +156,43 @@
         target_etypes = [target_etypes]
 
     if constructed_graph:
+        d_shuffled_nids = {} # to store shuffled nids by ntype to avoid different orders for the same ntype
         for target_e in target_etypes:
             num_edges = g.num_edges(target_e)
             g.edges[target_e].data['train_mask'] = th.full((num_edges,), False, dtype=th.bool)
             g.edges[target_e].data['val_mask'] = th.full((num_edges,), False, dtype=th.bool)
             g.edges[target_e].data['test_mask'] = th.full((num_edges,), False, dtype=th.bool)
-            g.edges[target_e].data['train_mask'][: int(num_edges * args.train_pct)] = True
-            g.edges[target_e].data['val_mask'][int(num_edges * args.train_pct): \
-                                               int(num_edges * (args.train_pct + args.val_pct))] = True
-            g.edges[target_e].data['test_mask'][int(num_edges * (args.train_pct + args.val_pct)): ] = True
+            if not args.inductive_split:
+                # Randomly split links
+                g.edges[target_e].data['train_mask'][: int(num_edges * args.train_pct)] = True
+                g.edges[target_e].data['val_mask'][int(num_edges * args.train_pct): \
+                                                int(num_edges * (args.train_pct + args.val_pct))] = True
+                g.edges[target_e].data['test_mask'][int(num_edges * (args.train_pct + args.val_pct)): ] = True
+            else:
+                # Inductive split for link prediction
+                # 1. split the head nodes u
+                utype, _, vtype = target_e
+                num_nodes = g.number_of_nodes(utype)
+                shuffled_index = d_shuffled_nids.get(utype,
+                                                     np.random.permutation(np.arange(num_nodes)))
+                if utype not in d_shuffled_nids:
+                    d_shuffled_nids[utype] = shuffled_index
+                train_u = shuffled_index[: int(num_nodes * args.train_pct)]
+                val_u = shuffled_index[int(num_nodes * args.train_pct): \
+                                        int(num_nodes * (args.train_pct + args.val_pct))]
+                test_u = shuffled_index[int(num_nodes * (args.train_pct + args.val_pct)): ]
+                # 2. find all out-edges for the sets of head nodes:
+                _, train_v, train_eids = g.out_edges(train_u, form='all', etype=target_e)
+                val_eids = g.out_edges(val_u, form='eid', etype=target_e)
+                test_eids = g.out_edges(test_u, form='eid', etype=target_e)
+                if utype == vtype:
+                    # we remove edges with tail nodes outside of the training set
+                    # this isn't necessary if head and tail are different types
+                    train_eids = train_eids[np.in1d(train_v, train_u)]
```

Comment:
What if there are some overlap between val and test?

Reply:
You're right. Removed val/test overlap in the next commit.